### PR TITLE
Add Snyk datasource

### PR DESCRIPTION
**Pull request by keshav-space**
closes https://github.com/nexB/vulnerablecode/issues/834

### CODE REVIEW

File: vulntotal/datasources/snyk.py [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+            "nuget": "nuget",
+            "pypi": "pip",
+            "rubygems": "rubygems",
+            # any purl.type not in supported_ecosystem shall implicitly be treated as unmanaged type
+            "unmanaged": "unmanaged",
+        }
+
+
+def generate_package_advisory_url(purl):
+    url_package_advisories = "https://security.snyk.io/package/{ecosystem}/{package}"
```
Comment (TG1999) [inline vulntotal/datasources/snyk.py:86]: Add docstring

File: vulntotal/datasources/snyk.py [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+        distro = purl.qualifiers["distro"]
+        purl_name = f"{distro}/{purl.name}"
+
+    return url_package_advisories.format(
+        ecosystem=supported_ecosystem[purl.type],
+        package=purl_name,
+    )
+
+
+def extract_html_json_advisories(package_advisories):
```
Comment (TG1999) [inline vulntotal/datasources/snyk.py:120]: Add docstring

File: vulntotal/datasources/snyk.py [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+                anchor = row.find(class_="vue--anchor")
+                ranges = row.find_all(
+                    "span", class_="vue--chip vulnerable-versions__chip vue--chip--default"
+                )
+                affected_versions = [vers.text.strip() for vers in ranges]
+                vulnerablity[anchor["href"].rsplit("/", 1)[-1]] = affected_versions
+    return vulnerablity
+
+
+def parse_html_advisory(advisory_html, snyk_id, affected) -> VendorData:
```
Comment (TG1999) [inline vulntotal/datasources/snyk.py:143]: Add docstring

File: vulntotal/datasources/snyk.py [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+        vulns_table = soup.find("tbody", class_="vue--table__tbody")
+        if vulns_table:
+            vulns_rows = vulns_table.find_all("tr", class_="vue--table__row")
+            for row in vulns_rows:
+                anchor = row.find(class_="vue--anchor")
+                ranges = row.find_all(
+                    "span", class_="vue--chip vulnerable-versions__chip vue--chip--default"
+                )
+                affected_versions = [vers.text.strip() for vers in ranges]
+                vulnerablity[anchor["href"].rsplit("/", 1)[-1]] = affected_versions
```
Comment (TG1999) [inline vulntotal/datasources/snyk.py:139]: Write a function for this and write some doctests for it

File: vulntotal/datasources/snyk.py [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+            fixed_versions = "".join(fixed[lower + 1 : upper]).split(",")
+    aliases.append(snyk_id)
+    return VendorData(
+        aliases=aliases,
+        affected_versions=affected,
+        fixed_versions=fixed_versions,
+    )
+
+
+def is_purl_in_affected(version, affected):
```
Comment (TG1999) [inline vulntotal/datasources/snyk.py:170]: Add docstring and doctest

**Review by TG1999 — CHANGES_REQUESTED** [review]

**Comment by pombredanne** [comment]
Merged in https://github.com/nexB/vulnerablecode/pull/1009

**Pull request closed without merging**